modules/videoio/src/cap_obsensor_capture.cpp

```diff
@@ -0,0 +1,145 @@
+    static const obsensor::StreamProfile depthProfile = {640, 480, 30, obsensor::FRAME_FORMAT_Y16};
+
+    streamChannelGroup_ = obsensor::getStreamChannelGroup(index);
+    if (!streamChannelGroup_.empty())
+    {
+        for (auto& channel : streamChannelGroup_)
+        {
+            auto streamType = channel->streamType();
+            switch (streamType)
+            {
+            case obsensor::OBSENSOR_STREAM_COLOR:
+                channel->start(colorProfile, [&](obsensor::Frame* frame) {
+                    std::unique_lock<std::mutex> lk(frameMutex_);
+                    colorFrame_ = Mat(1, frame->dataSize, CV_8UC1, frame->data).clone();
+                    });
+                break;
+            case obsensor::OBSENSOR_STREAM_DEPTH:
+                uint8_t data;
+                channel->setProperty(obsensor::DEPTH_TO_COLOR_ALIGN, &data, 1);
+                channel->start(depthProfile, [&](obsensor::Frame* frame) {
+                    std::unique_lock<std::mutex> lk(frameMutex_);
+                    depthFrame_ = Mat(frame->height, frame->width, CV_16UC1, frame->data, frame->width * 2).clone();
+                    });
+                uint32_t len;
+                memset(&camParam_, 0, sizeof(camParam_));
+                channel->getProperty(obsensor::CAMERA_PARAM, (uint8_t*)&camParam_, &len);
+                break;
+            default:
+                break;
+            }
+        }
+        isOpened_ = true;
+    }
+}
+
+bool VideoCapture_obsensor::grabFrame()
+{
+    std::unique_lock<std::mutex> lk(frameMutex_);
+
+    grabbedDepthFrame_ = depthFrame_;
+    grabbedColorFrame_ = colorFrame_;
+
+    depthFrame_.release();
+    colorFrame_.release();
+
+    return !grabbedDepthFrame_.empty() || !grabbedColorFrame_.empty();
+}
+
+bool VideoCapture_obsensor::retrieveFrame(int outputType, OutputArray frame)
+{
+    std::unique_lock<std::mutex> lk(frameMutex_);
+    switch (outputType)
+    {
+    case CAP_OBSENSOR_DEPTH_MAP:
+        if (!grabbedDepthFrame_.empty())
+        {
+            grabbedDepthFrame_.copyTo(frame);
+            grabbedDepthFrame_.release();
+            return true;
+        }
+        break;
+    case CAP_OBSENSOR_BGR_IMAGE:
+        if (!grabbedColorFrame_.empty())
+        {
+            auto mat = imdecode(grabbedColorFrame_, IMREAD_COLOR);
+            grabbedColorFrame_.release();
+
+            if (!mat.empty())
+            {
+                mat.copyTo(frame);
+                return true;
+            }
+        }
+        break;
+    default:
+        break;
+    }
+
+    return false;
+}
+
+double VideoCapture_obsensor::getProperty(int propIdx) const {
+    double rst = 0.0;
+    propIdx = propIdx & (~CAP_OBSENSOR_GENERATORS_MASK);
+    // int gen = propIdx & CAP_OBSENSOR_GENERATORS_MASK;
+    switch (propIdx) {
+    case CAP_PROP_OBSENSOR_INTRINSIC_FX:
+        rst = camParam_.p1[0] / (int)(camParam_.p1[2] * 2 / 640 + 0.5);
```

Comment:
the same expression is duplicated 4 times. Can we move `camParam_.p1[2]*2/640 + 0.5` outside of switch and give this scale some meaningful name?